---
title: "pdac scRNAseq"
author: "Marc Elosua-Bayes"
date: "4/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(purrr)
library(tibble)
library(ggplot2)
library(Matrix)
library(Seurat)
source("utils/bin.r")

tech <- "indrop"
tissue <- "pdac_itai_2"
dwn_smplng <- "both"
org <- "hs"

source("misc/paths_vrs.R")
```

## Introduction
In this R markdown document we will take a look at the scRNAseq data from [Integrating microarray-based spatial transcriptomics and single-cell RNA-seq reveals tissue architecture in pancreatic ductal adenocarcinomas](https://www.nature.com/articles/s41587-019-0392-8). It can be accessed under GEO accession number GSE111672.
This data was obtained using inDrop, between 139-145M paired-reads were  generated for each library (each one containing 2-2.5k cells) corresponding to ~58k paired-reads per cell. In this particular case we got in touch with the author's and obtained the annotated files.

Common parameters
```{r}
sample_dict <- list()
sample_dict[["GSM3036909"]] = "PDAC-A"
sample_dict[["GSM3036910"]] = "PDAC-A"
sample_dict[["GSM3405527"]] = "PDAC-A"
sample_dict[["GSM3405528"]] = "PDAC-A"
sample_dict[["GSM3405529"]] = "PDAC-A"
sample_dict[["GSM3405530"]] = "PDAC-A"
sample_dict[["GSM3405531"]] = "PDAC-B"
sample_dict[["GSM3405532"]] = "PDAC-B"
sample_dict[["GSM3405533"]] = "PDAC-B"
sample_dict[["GSM4100717"]] = "PDAC-C"
sample_dict[["GSM4100718"]] = "PDAC-C"
sample_dict[["GSM4100719"]] = "PDAC-C"
sample_dict[["GSM4100720"]] = "PDAC-C"


library(RColorBrewer)
n <- 60
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector  <-  unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))

pdacA <- c("GSM3036909", "GSM3036910", "GSM3405527", "GSM3405528", "GSM3405529", "GSM3405530")
pdacB <- c("GSM3405531", "GSM3405532", "GSM3405533")
pdacC <- c("GSM4100717", "GSM4100718", "GSM4100719", "GSM4100720")

```

## Load data
### scRNAseq filtered matrices
#### PDAC_A
```{r}
pdac_a_counts <- readr::read_tsv("analysis/pancreas_PDAC/data/Reuben_PDAC_A_B _annotation/PDAC-A-indrop-filtered.txt", col_names = FALSE)
colnames(pdac_a_counts) <- c("Genes", paste("cell_a", 2:ncol(pdac_a_counts), sep = "_"))

# Find duplicate genes
pdac_a_counts$Genes[duplicated(pdac_a_counts$Genes)]

# Remove duplicates based on Genes columns
pdac_a_counts <- pdac_a_counts[!duplicated(pdac_a_counts$Genes), ] %>% 
  tibble::column_to_rownames("Genes")

pdac_a_counts_mtrx <-   Matrix::Matrix(as.matrix(pdac_a_counts[-1, ]), sparse = TRUE)

annotation_a <- data.frame(t(pdac_a_counts[1, ]))
colnames(annotation_a) <- "annotation"
```

Create Seurat object
```{r}
pdac_A <- CreateSeuratObject(counts = pdac_a_counts_mtrx,
                                      project = "pdac_a", 
                                      assay = "RNA", 
                                      meta.data = annotation_a)
saveRDS(object = pdac_A,
        file = "analysis/pancreas_PDAC/data/PDAC-A_itai_joint.RDS")
pdac_A <- readRDS(file = "analysis/pancreas_PDAC/data/PDAC-A_itai_joint.RDS")
```

#### PDAC_B
```{r}
pdac_b_counts <- readr::read_tsv("analysis/pancreas_PDAC/data/Reuben_PDAC_A_B _annotation/PDAC-B-indrop-filtered.txt", col_names = FALSE)
colnames(pdac_b_counts) <- c("Genes", paste("cell_a", 2:ncol(pdac_b_counts), sep = "_"))

# Find duplicate genes
pdac_b_counts$Genes[duplicated(pdac_b_counts$Genes)]

# Remove duplicates based on Genes columns
pdac_b_counts <- pdac_b_counts[!duplicated(pdac_b_counts$Genes), ] %>% 
  tibble::column_to_rownames("Genes")

pdac_b_counts_mtrx <-   Matrix::Matrix(as.matrix(pdac_b_counts[-1, ]), sparse = TRUE)

annotation_b <- data.frame(t(pdac_b_counts[1, ]))
colnames(annotation_b) <- "annotation"
```

Create Seurat object
```{r}
pdac_B <- CreateSeuratObject(counts = pdac_b_counts_mtrx,
                                      project = "pdac_b",
                                      assay = "RNA",
                                      meta.data = annotation_b)
saveRDS(object = pdac_B, file = "analysis/pancreas_PDAC/data/PDAC-B_itai_joint.RDS")
pdac_B <- readRDS(file = "analysis/pancreas_PDAC/data/PDAC-B_itai_joint.RDS")

```

## Scale and normalize the data
### PDAC-A
```{r}
pdac_A <- Seurat::SCTransform(object = pdac_A)
# pdac_A <- Seurat::NormalizeData(object = pdac_A)
# pdac_A <- Seurat::FindVariableFeatures(pdac_A)
# pdac_A <- Seurat::ScaleData(pdac_A)
pdac_A <- Seurat::RunPCA(pdac_A, verbose = FALSE)
Seurat::ElbowPlot(pdac_A, ndims = 50)
```

From the elbow plot we can see that the elbow is around 12 so we will use the first 40 PC to proceed with the analysis.
```{r}
pdac_A <- Seurat::FindNeighbors(pdac_A,
                                 dims = 1:40)
pdac_A <- Seurat::FindClusters(pdac_A,
                                verbose = FALSE, 
                                resolution = c(1, 2, 3, 4, 5))
pdac_A <- Seurat::RunUMAP(pdac_A,
                           dims = 1:40)
```

Change names to match the original manuscript
```{r}
pdac_a_annot <- data.frame(
  annotation = sort(unique(as.character(pdac_A$annotation))),
  new_name = c("Acinar cells", "Cancer clone (TM4SF1)", "Cancer clone (S100A4)", 
               "High/Hypoxic ductal cells (APOL1)", "Centroacinar ductal cells",
               "Antigen-presenting ductal cells", 
               "Terminal ductal cells", "Endocrine cells", "Endothelial cells", 
               "Fibroblasts", "Macrophages M2", "Macrophages M1", "Mast cells",
               "mDCs", "mDCs", "Monocytes", "pDCs", "RBCs", "T cells & NK cells",
               "Tuft cells"))

new_annot <- data.frame(pdac_A@meta.data) %>% 
  left_join(pdac_a_annot, by = "annotation") %>% 
  pull(new_name) %>% 
  as.character()

pdac_A@meta.data[["annotation"]] <- new_annot
Idents(pdac_A) <- pdac_A$annotation
```

Save Seurat object
```{r}
saveRDS(object = pdac_A,
        file = "analysis/pancreas_PDAC/data/PDAC-A_itai_processed.RDS")
```

##### UMAP Author's Annotation
```{r eval = FALSE}
Idents(pdac_A) <- pdac_A$annotation
umap_pdac_a <- DimPlot(pdac_A, reduction = "umap", group.by = "annotation") +
  scale_color_manual(values = as.character(df_col[as.character(df_col$plt_name) %in% pdac_A$annotation, "col_ct"]))

ggpubr::ggexport(plotlist = list(umap_pdac_a),
                 filename = sprintf("%s/%s/Supplementary_Figure_QQQ_UMAP_PDAC-A.pdf",
                                    an_pdac,plt_dir),
                 width = 12,
                 height = 9,
                 res = 600)
```

### PDAC-B
```{r}
pdac_B <- Seurat::SCTransform(object = pdac_B)
# pdac_B <- Seurat::NormalizeData(object = pdac_B)
# pdac_B <- Seurat::FindVariableFeatures(pdac_B)
# pdac_B <- Seurat::ScaleData(pdac_A)
pdac_B <- Seurat::RunPCA(pdac_B, verbose = FALSE)
Seurat::ElbowPlot(pdac_B, ndims = 50)
```

From the elbow plot we can see that the elbow is around 12 so we will use the first 40 PC to proceed with the analysis.
```{r}
pdac_B <- Seurat::FindNeighbors(pdac_B,
                                 dims = 1:40)
pdac_B <- Seurat::FindClusters(pdac_B,
                                verbose = FALSE, 
                                resolution = c(1, 2, 3, 4, 5))
pdac_B <- Seurat::RunUMAP(pdac_B,
                           dims = 1:40)
```

Change annotation names
```{r}
pdac_b_annot <- data.frame(
  annotation = sort(unique(as.character(pdac_B$annotation))),
  new_name = c("Acinar cells", "Cancer clone (TM4SF1)", "Centroacinar ductal cells",
  "Antigen-presenting ductal cells", 
  "Terminal ductal cells", "Endocrine cells", 
  "Endothelial cells", "Macrophages", "Mast cells", "mDCs", "Monocytes", "RBCs",
  "Tuft cells"))

new_annot_b <- data.frame(pdac_B@meta.data) %>% 
  left_join(pdac_b_annot, by = "annotation") %>% 
  pull(new_name) %>% 
  as.character()

pdac_B@meta.data[["annotation"]] <- new_annot_b
Idents(pdac_B) <- pdac_B$annotation
```

Save object
```{r}
saveRDS(object = pdac_B, file = "analysis/pancreas_PDAC/data/PDAC-B_itai_processed.RDS")
```

##### UMAP Author's Annotation
```{r eval = FALSE}
Idents(pdac_B) <- pdac_B$annotation
umap_pdac_b <- DimPlot(pdac_B, reduction = "umap", group.by = "annotation") +
  scale_color_manual(values = as.character(df_col[as.character(df_col$plt_name) %in% pdac_B$annotation, "col_ct"]))

ggpubr::ggexport(plotlist = list(umap_pdac_b),
                 filename = sprintf("%s/%s/Supplementary_Figure_QQQ_UMAP_PDAC-B.pdf",
                                    an_pdac,plt_dir),
                 width = 12,
                 height = 9,
                 res = 600)

```

```{r}
QC1 <- QC_plots_fun(se_obj = pdac_B,
             count_thresh = ncounts,
             gene_thresh = ngenes,
             mt_thresh = mtcpct,
             vrs_names = c("nCount_RNA", "nFeature_RNA", "percent.mt", "orig.ident"))

QC2 <- QC_UMAP_fun(se_obj = pdac_B, 
                   vrs_names = c("nCount_RNA", "nFeature_RNA", "percent.mt", "orig.ident"),
                   mt_thresh = mtcpct,
                   gene_thresh = ngenes,
                   count_thresh = ncounts)

arr1 <- ggpubr::ggarrange(plotlist = QC1[1:3], nrow = 1, ncol = 3)
arr2 <- ggpubr::ggarrange(plotlist = list(QC2, QC1[[4]]), nrow = 1, ncol = 2)
arr_tot <- ggpubr::ggarrange(plotlist = list(arr1, arr2), nrow = 2)
```
