---
title: "pdac scRNAseq"
author: "Marc Elosua-Bayes"
date: "4/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(purrr)
library(tibble)
library(ggplot2)
library(Matrix)
library(Seurat)
source("utils/bin.r")

tech <- "indrop"
tissue <- "pdac_itai_2"
dwn_smplng <- "both"
org <- "hs"

source("misc/paths_vrs.R")
```

## Introduction
In this R markdown document we will take a look at the scRNAseq data from [Integrating microarray-based spatial transcriptomics and single-cell RNA-seq reveals tissue architecture in pancreatic ductal adenocarcinomas](https://www.nature.com/articles/s41587-019-0392-8). It can be accessed under GEO accession number GSE111672.
This data was obtained using inDrop, between 139-145M paired-reads were  generated for each library (each one containing 2-2.5k cells) corresponding to ~58k paired-reads per cell.

Common parameters
```{r}
sample_dict <- list()
sample_dict[["GSM3036909"]] = "PDAC-A"
sample_dict[["GSM3036910"]] = "PDAC-A"
sample_dict[["GSM3405527"]] = "PDAC-A"
sample_dict[["GSM3405528"]] = "PDAC-A"
sample_dict[["GSM3405529"]] = "PDAC-A"
sample_dict[["GSM3405530"]] = "PDAC-A"
sample_dict[["GSM3405531"]] = "PDAC-B"
sample_dict[["GSM3405532"]] = "PDAC-B"
sample_dict[["GSM3405533"]] = "PDAC-B"
sample_dict[["GSM4100717"]] = "PDAC-C"
sample_dict[["GSM4100718"]] = "PDAC-C"
sample_dict[["GSM4100719"]] = "PDAC-C"
sample_dict[["GSM4100720"]] = "PDAC-C"


library(RColorBrewer)
n <- 60
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector  <-  unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
```

## Load data
```{r eval = FALSE}
sample_geo <- c("GSM3036909", "GSM3036910", "GSM3405527", "GSM3405528", 
                "GSM3405529", "GSM3405530", "GSM3405531", "GSM3405532", 
                "GSM3405533", "GSM4100717", "GSM4100718", "GSM4100719", 
                "GSM4100720")

all_indrop <- list.files(path = "analysis/pancreas_PDAC/data", pattern = "indrop", recursive = TRUE)
z_indrop <- all_indrop[grepl(pattern = "^PDAC", x = all_indrop)] 

counts_sparse <- lapply(sample_geo, function(geo){
  sample <- z_indrop[grepl(pattern = geo, x = z_indrop)]
  
  geo_path <- sprintf("analysis/pancreas_PDAC/data/%s", sample)
  
  count_df <- readr::read_tsv(file = geo_path) %>% 
    tibble::column_to_rownames("Genes")
  colnames(count_df) <- paste(geo, colnames(count_df), sep = "-")
  count_df <- count_df %>% tibble::rownames_to_column("Genes")
  return(count_df)
  
}) %>% purrr::reduce(., function(...) dplyr::full_join(..., by = "Genes")) %>% 
  tibble::column_to_rownames("Genes") %>% 
  as.matrix()

counts_sparse[is.na(counts_sparse)] <- 0
counts_sparse <- Matrix(counts_sparse, sparse = TRUE)

pdac_se <- Seurat::CreateSeuratObject(counts = counts_sparse,
                                      project = "PDAC_itai", 
                                      min.cells = 1)
pdac_se[["percent.mt"]] <- PercentageFeatureSet(pdac_se, pattern = "^MT-")
saveRDS(object = pdac_se, 
        file = "analysis/pancreas_PDAC/data/pdac_itai_joint.RDS")

```

```{r eval = FALSE}
sample_A <- c("GSM3036909", "GSM3036910", "GSM3405527", "GSM3405528", 
                "GSM3405529", "GSM3405530")

all_indrop <- list.files(path = "analysis/pancreas_PDAC/data", pattern = "indrop", recursive = TRUE)
z_indrop <- all_indrop[grepl(pattern = "^PDAC", x = all_indrop)] 

counts_A <- lapply(sample_A, function(geo){
  sample <- z_indrop[grepl(pattern = geo, x = z_indrop)]
  
  geo_path <- sprintf("analysis/pancreas_PDAC/data/%s", sample)
  
  count_df <- readr::read_tsv(file = geo_path) %>% 
    tibble::column_to_rownames("Genes")
  colnames(count_df) <- paste(geo, colnames(count_df), sep = "-")
  count_df <- count_df %>% tibble::rownames_to_column("Genes")
  return(count_df)
  
}) %>% purrr::reduce(., function(...) dplyr::full_join(..., by = "Genes")) %>% 
  tibble::column_to_rownames("Genes") %>% 
  as.matrix()

counts_A[is.na(counts_A)] <- 0
counts_A <- Matrix(counts_A, sparse = TRUE)

pdac_A <- Seurat::CreateSeuratObject(counts = counts_A,
                                      project = "PDAC_A", 
                                      min.cells = 1)
pdac_A[["percent.mt"]] <- PercentageFeatureSet(pdac_A, pattern = "^MT-")
pdac_A[["percent.rr"]] <- PercentageFeatureSet(pdac_A, pattern = "^RP[A-Z][[:digit:]]")

saveRDS(object = pdac_A,
        file = "analysis/pancreas_PDAC/data/PDAC-A_itai_joint.RDS")

```

```{r eval = FALSE}
sample_B <- c("GSM3405531", "GSM3405532", "GSM3405533")

counts_B <- lapply(sample_B, function(geo){
  sample <- z_indrop[grepl(pattern = geo, x = z_indrop)]
  
  geo_path <- sprintf("analysis/pancreas_PDAC/data/%s", sample)
  
  count_df <- readr::read_tsv(file = geo_path) %>% 
    tibble::column_to_rownames("Genes")
  colnames(count_df) <- paste(geo, colnames(count_df), sep = "-")
  count_df <- count_df %>% tibble::rownames_to_column("Genes")
  return(count_df)
  
}) %>% purrr::reduce(., function(...) dplyr::full_join(..., by = "Genes")) %>% 
  tibble::column_to_rownames("Genes") %>% 
  as.matrix()

counts_B[is.na(counts_B)] <- 0
counts_B <- Matrix(counts_B, sparse = TRUE)

pdac_B <- Seurat::CreateSeuratObject(counts = counts_B,
                                      project = "PDAC_B", 
                                      min.cells = 1)
pdac_B[["percent.mt"]] <- PercentageFeatureSet(pdac_B, pattern = "^MT-")
pdac_B[["percent.rr"]] <- PercentageFeatureSet(pdac_B, pattern = "^RP[A-Z][[:digit:]]")

saveRDS(object = pdac_B,
        file = "analysis/pancreas_PDAC/data/PDAC-B_itai_joint.RDS")
```

```{r eval = FALSE}
sample_C <- c("GSM4100717", "GSM4100718", "GSM4100719", "GSM4100720")

counts_C <- lapply(sample_C, function(geo){
  sample <- z_indrop[grepl(pattern = geo, x = z_indrop)]
  
  geo_path <- sprintf("analysis/pancreas_PDAC/data/%s", sample)
  
  count_df <- readr::read_tsv(file = geo_path) %>% 
    tibble::column_to_rownames("Genes")
  colnames(count_df) <- paste(geo, colnames(count_df), sep = "-")
  count_df <- count_df %>% tibble::rownames_to_column("Genes")
  return(count_df)
  
}) %>% purrr::reduce(., function(...) dplyr::full_join(..., by = "Genes")) %>% 
  tibble::column_to_rownames("Genes") %>% 
  as.matrix()

counts_C[is.na(counts_C)] <- 0
counts_C <- Matrix(counts_C, sparse = TRUE)

pdac_C <- Seurat::CreateSeuratObject(counts = counts_C,
                                      project = "PDAC_C", 
                                      min.cells = 1)
pdac_C[["percent.mt"]] <- PercentageFeatureSet(pdac_C, pattern = "^MT-")
pdac_C[["percent.rr"]] <- PercentageFeatureSet(pdac_C, pattern = "^RP[A-Z][[:digit:]]")

saveRDS(object = pdac_C,
        file = "analysis/pancreas_PDAC/data/PDAC-C_itai_joint.RDS")
```


```{r eval = FALSE}
pdacA <- c("GSM3036909", "GSM3036910", "GSM3405527", "GSM3405528", "GSM3405529", "GSM3405530")
pdacB <- c("GSM3405531", "GSM3405532", "GSM3405533")
pdacC <- c("GSM4100717", "GSM4100718", "GSM4100719", "GSM4100720")


pdac_se <- readRDS(file = "analysis/pancreas_PDAC/data/pdac_itai_joint.RDS")
pdac_se[["percent.mt"]] <- PercentageFeatureSet(pdac_se, pattern = "^MT-")

pdac_se@meta.data <- pdac_se@meta.data %>% 
  tidyr::separate(data = ., col = orig.ident, into = c("geo", NA)) %>% 
  dplyr::mutate(sample = if_else(geo %in% pdacA, "PDAC-A", 
                          if_else(geo %in% pdacB, "PDAC-B", "PDAC-C")))

metaA <- pdac_se@meta.data[pdac_se$sample == "PDAC-A", ]
metaB <- pdac_se@meta.data[pdac_se$sample == "PDAC-B", ]
metaC <- pdac_se@meta.data[pdac_se$sample == "PDAC-C", ]

pdac_A <- pdac_se[, Idents(pdac_se) == "PDAC-A"]
pdac_A@meta.data <- metaA
pdac_B <- pdac_se[, pdac_se$sample == "PDAC-B"]
pdac_B@meta.data <- metaB
pdac_C <- pdac_se[, pdac_se$sample == "PDAC-C"]
pdac_C@meta.data <- metaC
```

### scRNAseq filtered matrices
#### PDAC_A
```{r}
pdac_a_counts <- readr::read_tsv("analysis/pancreas_PDAC/data/Reuben_PDAC_A_B _annotation/PDAC-A-indrop-filtered.txt", col_names = FALSE)
colnames(pdac_a_counts) <- c("Genes", paste("cell_a", 2:ncol(pdac_a_counts), sep = "_"))

# Find duplicate genes
pdac_a_counts$Genes[duplicated(pdac_a_counts$Genes)]

# Remove duplicates based on Genes columns
pdac_a_counts <- pdac_a_counts[!duplicated(pdac_a_counts$Genes), ] %>% 
  tibble::column_to_rownames("Genes")

pdac_a_counts_mtrx <-   Matrix::Matrix(as.matrix(pdac_a_counts[-1, ]), sparse = TRUE)

annotation_a <- data.frame(t(pdac_a_counts[1, ]))
colnames(annotation_a) <- "annotation"
```

Create Seurat object
```{r}
pdac_A <- CreateSeuratObject(counts = pdac_a_counts_mtrx,
                                      project = "pdac_a", 
                                      assay = "RNA", 
                                      meta.data = annotation_a)
saveRDS(object = pdac_A,
        file = "analysis/pancreas_PDAC/data/PDAC-A_itai_joint.RDS")
pdac_A <- readRDS(file = "analysis/pancreas_PDAC/data/PDAC-A_itai_joint.RDS")
```

#### PDAC_B
```{r}
pdac_b_counts <- readr::read_tsv("analysis/pancreas_PDAC/data/Reuben_PDAC_A_B _annotation/PDAC-B-indrop-filtered.txt", col_names = FALSE)
colnames(pdac_b_counts) <- c("Genes", paste("cell_a", 2:ncol(pdac_b_counts), sep = "_"))

# Find duplicate genes
pdac_b_counts$Genes[duplicated(pdac_b_counts$Genes)]

# Remove duplicates based on Genes columns
pdac_b_counts <- pdac_b_counts[!duplicated(pdac_b_counts$Genes), ] %>% 
  tibble::column_to_rownames("Genes")

pdac_b_counts_mtrx <-   Matrix::Matrix(as.matrix(pdac_b_counts[-1, ]), sparse = TRUE)

annotation_b <- data.frame(t(pdac_b_counts[1, ]))
colnames(annotation_b) <- "annotation"
```

Create Seurat object
```{r}
pdac_B <- CreateSeuratObject(counts = pdac_b_counts_mtrx,
                                      project = "pdac_b",
                                      assay = "RNA",
                                      meta.data = annotation_b)
saveRDS(object = pdac_B, file = "analysis/pancreas_PDAC/data/PDAC-B_itai_joint.RDS")
pdac_B <- readRDS(file = "analysis/pancreas_PDAC/data/PDAC-B_itai_joint.RDS")

```

#### Load data joint
```{r}
# pdac_A <- readRDS(file = "analysis/pancreas_PDAC/data/PDAC-A_itai_joint.RDS")
# pdac_B <- readRDS(file = "analysis/pancreas_PDAC/data/PDAC-B_itai_joint.RDS")
# pdac_C <- readRDS(file = "analysis/pancreas_PDAC/data/PDAC-C_itai_joint.RDS")
```

### QC of data
We will start by applying the QC they did in the paper: 
*Only single-cell transcriptomes with ≥ 1000 UMIs, ≤ 20% mitochondrial transcripts and ≤ 30% ribosomal transcripts were kept, leaving 1926 cells for the PDAC-A dataset, and 1733 cells for the PDAC-B dataset.*
#### PDAC-A
In Itai's paper after filtering by these parameters there should be 1926 and 1733 cells in PDAC-A and PDAC-B but here we end up with `ncol(pdac_A)` and `ncol(pdac_B)` respectively. This could be due to the definition of ribosomal genes used here: ^RP[A-Z][[:digit:]]
```{r eval = FALSE}
ncounts <- 1000; mtcpct <- 20; rrpct <- 30; ngenes <- 0
```

```{r eval = FALSE}
pdac_A <- subset(x = pdac_A,
                 subset = nCount_RNA >= ncounts & 
                          percent.mt <= mtcpct & 
                          percent.rr <= rrpct)
pdac_B <- subset(x = pdac_B, 
                 subset = nCount_RNA >= ncounts & 
                          percent.mt <= mtcpct & 
                          percent.rr <= rrpct)
pdac_C <- subset(x = pdac_C, 
                 subset = nCount_RNA >= ncounts & 
                          percent.mt <= mtcpct & 
                          percent.rr <= rrpct)
```

```{r eval = FALSE}
VlnPlot(pdac_A, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rr"), ncol = 3)

plot1 <- FeatureScatter(pdac_A, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pdac_A, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

Scale and normalize the data
```{r}
pdac_A <- Seurat::SCTransform(object = pdac_A)
# pdac_A <- Seurat::NormalizeData(object = pdac_A)
# pdac_A <- Seurat::FindVariableFeatures(pdac_A)
# pdac_A <- Seurat::ScaleData(pdac_A)
pdac_A <- Seurat::RunPCA(pdac_A, verbose = FALSE)
Seurat::ElbowPlot(pdac_A, ndims = 50)
```

From the elbow plot we can see that the elbow is around 12 so we will use the first 40 PC to proceed with the analysis.
```{r}
pdac_A <- Seurat::FindNeighbors(pdac_A,
                                 dims = 1:40)
pdac_A <- Seurat::FindClusters(pdac_A,
                                verbose = FALSE, 
                                resolution = c(1, 2, 3, 4, 5))
pdac_A <- Seurat::RunUMAP(pdac_A,
                           dims = 1:40)
```

Marker genes to annotate specific cell types
```{r eval = FALSE}
Idents(pdac_A) <- pdac_A$annotation
markers_pdacA <- Seurat::FindAllMarkers(object = pdac_A, 
                                        verbose = TRUE, 
                                        only.pos = TRUE,
                                        assay = "SCT",
                                        slot = "data")

Seurat::FindMarkers(object = pdac_A,
                    ident.1 = "Macrophages B",
                    ident.2 = "Macrophages A",
                    verbose = TRUE, 
                    only.pos = TRUE,
                    assay = "SCT",
                    slot = "data") %>% 
  tibble::rownames_to_column("Gene") %>% 
  dplyr::filter(Gene %in% c("CD163", "Ms4A4A", "IL1B"))
```

Spatial Feature plot to look at "CD163", "Ms4A4A", "IL1B" to differentiate between both macrophage populations
```{r eval = FALSE}
fplt <- FeaturePlot(pdac_A, features = c("CD163", "Ms4A4A", "IL1B"))
HoverLocator(plot = fplt, information = FetchData(pdac_A, vars = c("ident")))
```

Macrophages A are expressing CD163 & Ms4A4A - M2
Macrophages B are expressing IL1B - M1


Compare both mDC populations
```{r eval = FALSE}
fplt <- FeaturePlot(pdac_A, 
                    features = c("CFB", "C4A", "C1S" , "C4B", "CFH"))

HoverLocator(plot = fplt, information = FetchData(pdac_A, vars = c("ident")))
```

Cannot observe the differnces in the mDC populations, thus will consider them as one.

Change names to match the original manuscript
```{r}
pdac_a_annot <- data.frame(
  annotation = sort(unique(as.character(pdac_A$annotation))),
  new_name = c("Acinar cells", "Cancer clone (TM4SF1)", "Cancer clone (S100A4)", 
               "High/Hypoxic ductal cells (APOL1)", "Centroacinar ductal cells",
               "Antigen-presenting ductal cells", 
               "Terminal ductal cells", "Endocrine cells", "Endothelial cells", 
               "Fibroblasts", "Macrophages M2", "Macrophages M1", "Mast cells",
               "mDCs", "mDCs", "Monocytes", "pDCs", "RBCs", "T cells & NK cells",
               "Tuft cells"))

new_annot <- data.frame(pdac_A@meta.data) %>% 
  left_join(pdac_a_annot, by = "annotation") %>% 
  pull(new_name) %>% 
  as.character()

pdac_A@meta.data[["annotation"]] <- new_annot
Idents(pdac_A) <- pdac_A$annotation
```

Save Seurat object
```{r}
saveRDS(object = pdac_A,
        file = "analysis/pancreas_PDAC/data/PDAC-A_itai_processed.RDS")
```

We're gonna check for batch effects 
```{r eval = FALSE}
table(pdac_A$orig.ident)
DimPlot(pdac_A, reduction = "umap", group.by = "orig.ident")
DimPlot(pdac_A, reduction = "umap", group.by = "SCT_snn_res.0.8")
DimPlot(pdac_A, reduction = "umap", group.by = "SCT_snn_res.0.5")
DimPlot(pdac_A, reduction = "umap", group.by = "SCT_snn_res.1")
DimPlot(pdac_A, reduction = "umap", group.by = "seurat_clusters")
```

```{r eval = FALSE}
QC1 <- QC_plots_fun(se_obj = pdac_A,
             count_thresh = ncounts,
             gene_thresh = ngenes,
             mt_thresh = mtcpct,
             vrs_names = c("nCount_RNA", "nFeature_RNA", "percent.mt", "orig.ident"))

QC2 <- QC_UMAP_fun(se_obj = pdac_A, 
                   vrs_names = c("nCount_RNA", "nFeature_RNA", "percent.mt", "orig.ident"),
                   mt_thresh = mtcpct,
                   gene_thresh = ngenes,
                   count_thresh = ncounts)

arr1 <- ggpubr::ggarrange(plotlist = QC1[1:3], nrow = 1, ncol = 3)
arr2 <- ggpubr::ggarrange(plotlist = list(QC2, QC1[[4]]), nrow = 1, ncol = 2)
arr_tot <- ggpubr::ggarrange(plotlist = list(arr1, arr2), nrow = 2)
```

#### PDAC-B
```{r eval = FALSE}
VlnPlot(pdac_B, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "percent.rr"), ncol = 3)

plot1 <- FeatureScatter(pdac_B, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pdac_B, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

Scale and normalize the data
```{r}
pdac_B <- Seurat::SCTransform(object = pdac_B)
# pdac_B <- Seurat::NormalizeData(object = pdac_B)
# pdac_B <- Seurat::FindVariableFeatures(pdac_B)
# pdac_B <- Seurat::ScaleData(pdac_A)
pdac_B <- Seurat::RunPCA(pdac_B, verbose = FALSE)
Seurat::ElbowPlot(pdac_B, ndims = 50)
```

From the elbow plot we can see that the elbow is around 12 so we will use the first 40 PC to proceed with the analysis.
```{r}
pdac_B <- Seurat::FindNeighbors(pdac_B,
                                 dims = 1:40)
pdac_B <- Seurat::FindClusters(pdac_B,
                                verbose = FALSE, 
                                resolution = c(1, 2, 3, 4, 5))
pdac_B <- Seurat::RunUMAP(pdac_B,
                           dims = 1:40)
```

Change annotation names
```{r}
pdac_b_annot <- data.frame(
  annotation = sort(unique(as.character(pdac_B$annotation))),
  new_name = c("Acinar cells", "Cancer clone (TM4SF1)", "Centroacinar ductal cells",
  "Antigen-presenting ductal cells", 
  "Terminal ductal cells", "Endocrine cells", 
  "Endothelial cells", "Macrophages", "Mast cells", "mDCs", "Monocytes", "RBCs",
  "Tuft cells"))

new_annot_b <- data.frame(pdac_B@meta.data) %>% 
  left_join(pdac_b_annot, by = "annotation") %>% 
  pull(new_name) %>% 
  as.character()

pdac_B@meta.data[["annotation"]] <- new_annot_b
Idents(pdac_B) <- pdac_B$annotation
```

Save object
```{r}
saveRDS(object = pdac_B, file = "analysis/pancreas_PDAC/data/PDAC-B_itai_processed.RDS")
```


We're gonna check for batch effects 
```{r eval = FALSE}
table(pdac_B$orig.ident)
DimPlot(pdac_B, reduction = "umap", group.by = "orig.ident")
DimPlot(pdac_B, reduction = "umap", group.by = "SCT_snn_res.0.8")
DimPlot(pdac_B, reduction = "umap", group.by = "SCT_snn_res.0.5")
DimPlot(pdac_B, reduction = "umap", group.by = "SCT_snn_res.1")
DimPlot(pdac_B, reduction = "umap", group.by = "seurat_clusters")
```

##### UMAP Reuben Annotation
```{r eval = FALSE}
Idents(pdac_B) <- pdac_B$annotation
umap_pdac_b <- DimPlot(pdac_B, reduction = "umap", group.by = "annotation") +
  scale_color_manual(values = as.character(df_col[as.character(df_col$plt_name) %in% pdac_B$annotation, "col_ct"]))

ggpubr::ggexport(plotlist = list(umap_pdac_b),
                 filename = sprintf("%s/%s/Supplementary_Figure_QQQ_UMAP_PDAC-B.pdf",
                                    an_pdac,plt_dir),
                 width = 12,
                 height = 9,
                 res = 600)

```

```{r}
QC1 <- QC_plots_fun(se_obj = pdac_B,
             count_thresh = ncounts,
             gene_thresh = ngenes,
             mt_thresh = mtcpct,
             vrs_names = c("nCount_RNA", "nFeature_RNA", "percent.mt", "orig.ident"))

QC2 <- QC_UMAP_fun(se_obj = pdac_B, 
                   vrs_names = c("nCount_RNA", "nFeature_RNA", "percent.mt", "orig.ident"),
                   mt_thresh = mtcpct,
                   gene_thresh = ngenes,
                   count_thresh = ncounts)

arr1 <- ggpubr::ggarrange(plotlist = QC1[1:3], nrow = 1, ncol = 3)
arr2 <- ggpubr::ggarrange(plotlist = list(QC2, QC1[[4]]), nrow = 1, ncol = 2)
arr_tot <- ggpubr::ggarrange(plotlist = list(arr1, arr2), nrow = 2)
```

### Annotate clusters
#### PDAC-A
```{r}
Seurat::Idents(object = pdac_A) <- pdac_A@meta.data[, "SCT_snn_res.5"]
markers_pdacA <- Seurat::FindAllMarkers(object = pdac_A, 
                                        verbose = TRUE, 
                                        only.pos = TRUE,
                                        assay = "SCT",
                                        slot = "data")
```

```{r}
Seurat::Idents(object = pdac_B) <- pdac_B@meta.data[, "SCT_snn_res.5"]
markers_pdacB <- Seurat::FindAllMarkers(object = pdac_B, 
                                        verbose = TRUE, 
                                        only.pos = TRUE,
                                        assay = "SCT",
                                        slot = "data")
```

**Malignant Ductal Cells**
The two PDAC-A cancer clusters – henceforth, cancer clusters 1 and 2 – were enriched for expression of either **TM4SF1** (cluster 1) or **S100A4** (cluster 2), while the PDAC-B cancer cluster was only enriched for **TM4SF1** expression (Fig. S2b), suggesting similarity between PDAC-A cancer cluster 1 and the PDAC-B cancer cluster.

**Non-Malignant Ductal subpopulation**
From both tumor samples, a majority of the scRNA-Seq cells consisted of KRT19 expressing ductal cells (Fig. 1b, S2b), one of the two main constituent cell types of the pancreatic exocrine system36. We identified a total of four ductal subpopulations: a ductal population expressing APOL1 and hypoxia-response related genes including ERO1A37 and CA938, a terminal ductal population expressing TFF1, TFF2, and TFF339, a centroacinar ductal population expressing CRISP3 and CFTR39, and antigen-presenting ductal cells expressing major histocompatibility complex (MHC) class II genes CD74, HLA-DPA1, HLA-DQA2, HLA-DRA, HLA-DRB1, HLA-DRB5, and complement pathway components C1S, C4A, C4B, CFB, and CFH (Fig. 3a-d). Although MHC class II molecules are primarily expressed on the surface of professional antigen-presenting cells (B-cells, macrophages, dendritic cells), epithelial cells in the liver, gastrointestinal and respiratory tracts are known to express MHC class II40,41. Because of their presence in the tumor, it is likely that these non-professional antigen-presenting ductal cells play a role in modulating the inflammatory response within the microenvironment by promoting T cell activation40,41. When we performed double IF of archival FFPE patient tissue, we found co-localization of subpopulation markers with the duct marker KRT19, confirming the presence of these ductal cell subpopulations (Fig. 3e-h, Fig. S6).

**Macrophages**
We also found that the PDAC-A macrophages comprised two subpopulations, one of which expressed IL1B, corresponding to an inflammatory M1 state,42,43 and the other resembling the M2 alternatively activated state based on its expression of CD163 and MS4A4A44,45

**Dendritic Cells**
We also found two subpopulations of dendritic cells, one of which expressed higher levels of complement pathway genes and MHC class II

We are going to use Paula's curated annotation list and add marker genes for non-present cell types or subtypes
```{r}
library(matchSCore2)
marker_ls <- readRDS(file = "my_curated_gene_list_human.rds")
marker_ls[["General Ductal cells"]] <- c("KRT19", "PROM1")
marker_ls[["Malignant Ductal Cells 1"]] <- c("TM4SF1")
marker_ls[["Malignant Ductal Cells 2"]] <- c("S100A4")
marker_ls[["Ductal subpopulation 1"]] <- c("APOL1", "ERO1A", "CA9")
marker_ls[["Terminal ductal subpopulation"]] <- c("TFF1", "TFF2", "TFF3")
marker_ls[["Centroacinar ductal population"]] <- c("CRISP3", "CFTR")
marker_ls[["Antigen-presenting ductal cells"]] <- c("CD74",  "HLA-DPA1", 
                                                    "HLA-DQA2", "HLA-DRA", 
                                                    "HLA-DRB1")
marker_ls[["MacrophagesI inflammatory M1"]] <- c("IL1B")
marker_ls[["MacrophagesII alternatively activated"]] <- c("MS4A4A", "CD163")
marker_ls[["MacrophagesI M1 (cell marker)"]] <- as.character(expression(CCR7, 
                                  CD127, CD16, CD32, CD62, CD64, CD80, CD86, 
                                  CXCL10, IL15R, IL17R, IL1R1, IL2R, TLR2, TLR4))
marker_ls[["MacrophagesI M2 (cell marker)"]] <- as.character(expression(CD163, 
                                  CD206, CD23, CD36, IL1, CD209, DCIR, CLACSF13,
                                  FIZZ1, M60, CD184, TRAI))
marker_ls[["Red blood cell"]] <- c("GlyA")
marker_ls[["PDAC fibroblasts"]] <- c("COL11A1")
marker_ls[["Normal fibroblasts"]] <- as.character(expression(CD105, CD11b, CD19,
                                  CD34, CD44, CD45, CD73, CD90, HLA-DR))
marker_ls[["Acinar cells"]] <- as.character(expression(AADAC, AKR1B10, AKR1C1, 
                                AKR1C2, AKR1C3, AKR7A3, ALB, ALDOB, AMY2A, 
                                AMY2B, AMYP1, ANPEP, AQP8, AZGP1, BCAT1, 
                                C15orf48, C4B, CA12, CBS, CEL, CELA2A, CELA3A, 
                                CELA3B, CLPS, CPA1, CPA2, CPB1, CTRB1, CTRB2, 
                                CTRC, CTRL, CUZD1, CXCL17, DPEP1, DUOX2, DUOXA2,
                                ECE2, EFHD2, ERP27, FAM129A, FGL1, GALNT7, GNL3,
                                GP2, GPT2, GSTA1, GSTA2, HAMP, IL32, KCTD14, 
                                KLK1, LGALS2, LYZ, MAT1A, MGST1, MUC1, NFKBIA, 
                                NUPR1, PDIA2, PLA2G1B, PLTP, PNLIP, PNLIPRP1, 
                                PNLIPRP2, PPIF, PRSS1, PRSS2, PRSS3, PRSS3P2, 
                                PSAT1, PTGR1, RARRES2, REG1A, REG1B, REG3A, 
                                REG3G, RNASE1, RP11_520H11.1, RP6_149D17.1, 
                                SDC4, SERPINA4, SERPINI2, SGK1, SLC25A37, 
                                SLC38A5, SLC39A5, SLC43A1, SOD2, SPINK1, SYCN, 
                                TIFA, TMEM97, TMPRSS2, TPD52L1, TPST2, TRIB2, 
                                U66061.39, UGDH, YBX3, ZDHHC9))

marker_ls[["conventional dendritic cell"]] <- as.character(expression(CD11c, 
                                CD135, CD14, CD141, CD172a, CD197, CD1b, CD1c, 
                                CD205, CD206, CD207, CD273, CD282, CD284, CD289,
                                CD369, CD370, CD40, CD49d, CLEC4C, IL3RA, NRP1))

marker_ls[["Plasmacytoid dendritic cell"]] <- as.character(expression(CD123, GZMB))


marker_ls[["beta cell"]] <- as.character(expression(BMP-5, CDKN1C, CRTR1, DLK1,
                                NPTX2, PACAP, PDX1, PFKFB2, PIR, SIX2, SIX3, 
                                SMAD9, SYT13, TGFBR3))
marker_ls[["pancreatic islet"]] <- as.character(expression(ABI2, AC009041.1, 
                                ADCYAP1, AL117190.2, ALCAM, ALOX5, AP3B1, APCDD1L,
                                APLP2, ATP2A3, BMP5, C1orf127, C21orf119, CADM1,
                                CAPN13, CASR, CD72, CDH18, CDK5R1, CGRRF1, CNTN4,
                                CSDE1, CTSV, CYP2U1, CYYR1, DLK1, EDARADD, 
                                EFCAB14, ELMO1, ELOVL4, ENPP4, ENTPD3, ERO1LB, 
                                FAM222A, FAM76A, FFAR1, G6PC2, GCC1, GNAS, GPM6A,
                                GPSM1, GREM1, GUSB, GYG1, HSPD1, IAPP, IGF2, INS,
                                KLHDC8A, LGMN, LRRTM3, MAFA, MRAP2, MTSS1, MXRA7,
                                NECAB1, NKX6-1, NMRK1, NPTX2, NTPCR, OXLD1, 
                                PCSK1, PFKFB2, PIGV, PIK3IP1, RAPH1, RASA4, 
                                RASA4B, RASGRF1, RBP4, RGS16, RNF130, ROBO2, 
                                RRAGB, RRNAD1, SALL2, SCD5, SEMA5A, SFTPA1, SGCB,
                                SGK3, SIAH1, SLC25A34, SLC30A8, SLC43A2, SOCS2, 
                                SORL1, SRD5A1, SUSD4, SYT13, TBCC, TDRP, TERF2IP,
                                TMEM37, TSPAN1, VAT1L, WNT4, WSCD2, ZNF441, ZNF776))

marker_ls[["MAST"]] <- as.character(expression(CD117, CD203c, CD25, KIT, SLC18A2))
marker_ls[["Monocyte"]] <- as.character(expression(CD14, CD11b, CCR2, CD16, CD141, CD11c, HLA-DR, CCR7))
marker_ls[["Endocrine"]] <- as.character(expression(CD99, DISP2, LRP11, SEZ6L2, SLC30A8))
marker_ls[["Endothelial"]] <- as.character(expression(CD31, CD34, CD146, CD105, VWF))
marker_ls[["Erythrocytes"]] <- as.character(expression(CD235A, CD45))
marker_ls[["Fibroblasts (inflamatory)"]] <- as.character(expression(IL-6, CD90, CD10, CD29, CD105))
```

We use matchscore to annotate the different clusters:
https://github.com/elimereu/matchSCore2
```{r}
markers_pdacA_ls <- lapply(unique(markers_pdacA$cluster), function(i) {
  markers_pdacA[markers_pdacA$cluster == i, "gene"]
})

ms <- matchSCore2::matchSCore2(gene_cl.ref = marker_ls, 
                               gene_cl.obs = markers_pdacA_ls, 
                               ylab = "gene_list", 
                               xlab = "clusters")
ms
```

Together with matchsocre we are going to look at gene expression in the clusters
```{r}
ref_plt <- DimPlot(pdac_A, reduction = "umap", group.by = "SCT_snn_res.5", label = TRUE)
ref_plt_1617 <- DimPlot(pdac_A[, pdac_A$SCT_snn_res.5 %in% 16:17], reduction = "umap", group.by = "SCT_snn_res.5", label = TRUE)
ref_plt_2924 <- DimPlot(pdac_A[, pdac_A$SCT_snn_res.5 %in% c(29, 24)], reduction = "umap", group.by = "SCT_snn_res.5", label = TRUE)
ref_plt_23 <- DimPlot(pdac_A[, pdac_A$SCT_snn_res.5 %in% 23], reduction = "umap", group.by = "SCT_snn_res.5", label = TRUE)
HoverLocator(plot = ref_plt, information = FetchData(pdac_A, vars = c("SCT_snn_res.5")))


Seurat::FeaturePlot(object = pdac_A,
                    features = c("TM4SF1", # Cancer subtype marker mutually exclusive with S100A4
                                 "S100A4", # Cancer subtype marker mutually exclusive with TM4SF1
                                 "KRT19" # Ubiquitous in malignant and non-malignant ductal cells 
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_A,
                    features = c("APOL1", # Hypoxic ductal subpopulation 1
                                 "ERO1A", # Hypoxic ductal subpopulation 1
                                 "CA9" # Hypoxic ductal subpopulation 1
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_A,
                    features = c("TFF1", # Terminal ductal subpopulation
                                 "TFF2", # Terminal ductal subpopulation
                                 "TFF3" # Terminal ductal subpopulation
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_A,
                    features = c("CRISP3", # centroacinar ductal population
                                 "CFTR", # centroacinar ductal population
                                 "TFF1",
                                 "AQP3"
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_A,
                    features = c("CD74", # antigen-presenting ductal cells
                                 "HLA-DPA1", # antigen-presenting ductal cells
                                 "HLA-DQA2", # antigen-presenting ductal cells
                                 "HLA-DRA", # antigen-presenting ductal cells
                                 "HLA-DRB1", # antigen-presenting ductal cells
                                 "C1S", "C4A", "C4B", "CFB", "CFH"
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_A[, pdac_A$SCT_snn_res.5 %in% 16:17],
                    features = c("IL1B", # MacrophagesI, inflammatory M1 stat
                                 "CD163", # MacrophagesII alternatively activated
                                 "MS4A4A", # MacrophagesII alternatively activated
                                 "CD68", # General Myeloyd marker
                                 "C1QA" # Macrophage specific
                                 )) + ref_plt_1617

### MAST ###
Seurat::FeaturePlot(object = pdac_A, 
                    features = c("CD117", "CD203c", "CD25", "KIT", "SLC18A2")) + ref_plt

### Monocyte ###
Seurat::FeaturePlot(object = pdac_A[, pdac_A$SCT_snn_res.5 %in% 16:17], 
                    features = c("CD14", "CCR2", "CCR7", "CD68")) + ref_plt_1617

### DC ###
Seurat::FeaturePlot(object = pdac_A, 
                    features = c("CD40", "IL3RA", "NRP1", "CD141", "BDCA3")) + ref_plt
# "CD11c", "CD135", "CD141", "CD172a", "CD197", "CD1b", "CD1c", "CD205", "CD206",  "CD273", "CD282",  "CD284", "CD289", "CD369", "CD370", "CD49d", "CD207"

### pDCs ###
Seurat::FeaturePlot(object = pdac_A[, pdac_A$SCT_snn_res.5 %in% c(29, 24)], 
                    features = c("CD123", "GZMB", "BDCA2", "NRP1", "CLEC4C")) + ref_plt_2924
Seurat::FeaturePlot(object = pdac_A, 
                    features = c("CD123", "GZMB", "BDCA2", "NRP1", "CLEC4C")) + ref_plt

### NK ###
Seurat::FeaturePlot(object = pdac_A[, pdac_A$SCT_snn_res.5 %in% c(29, 24)], 
                    features = marker_ls[["NK"]]) + ref_plt_2924


### CD8 T cells ###
Seurat::FeaturePlot(object = pdac_A[, pdac_A$SCT_snn_res.5 %in% c(29, 24)], 
                    features = marker_ls[["CD8+ T cells"]]) + ref_plt_2924

### CD4 T cells ###
Seurat::FeaturePlot(object = pdac_A[, pdac_A$SCT_snn_res.5 %in% c(29, 24)], 
                    features = marker_ls[["CD4+ T cells"]]) + ref_plt_2924

### Pancreatic islet cells ###
Seurat::FeaturePlot(object = pdac_A, 
                    features = c(marker_ls[["pancreatic islet"]][1:10], "AZGP1")) + ref_plt

### Endothelial ###
Seurat::FeaturePlot(object = pdac_A, 
                    features = c("CD31", "CD34", "CD146", "CD105", "VWF")) + ref_plt

### Endocrine ###
Seurat::FeaturePlot(object = pdac_A, 
                    features = c("CD99", "SEZ6L2", "DNER")) + ref_plt
# c("CD99", "DISP2", "LRP11", "SEZ6L2", "SLC30A8", "DDR1", "DNER")
### Erythrocytes ###
Seurat::FeaturePlot(object = pdac_A, 
                    features = c("ALAS2", "HBA1", "HBB", "HBG1")) + ref_plt

### tufts cells ###
Seurat::FeaturePlot(object = pdac_A, 
                    features = c("ACADSB", "ACSL4", "ADA", "ADAM10", "AEN", 
                                 "AFAP1", "AHCY", "ALPL", "ANXA1", "ANXA4", 
                                 "ASCL1", "ASCL2", "ATP11C", "AZGP1")) + ref_plt

DimPlot(pdac_A, reduction = "umap", group.by = "SCT_snn_res.5", label = TRUE) + theme_bw()
```

If we look at cluster 23 we don't hve it only in 1 place so we're going to annotate the different clusters separately
```{r}
cl23_sel <- pdac_A@meta.data %>% 
  data.frame() %>% 
  tibble::rownames_to_column("ID") %>% 
  left_join(data.frame(pdac_A@reductions$umap@cell.embeddings) %>% rownames_to_column("ID"), by = "ID")

annot_23 <- cl23_sel %>%
  mutate(
    pDCs = if_else(UMAP_1 > 15 & UMAP_2 > 0, TRUE, FALSE),
    endothelial = if_else(UMAP_1 > -0.2 & UMAP_1 < 0.5 & UMAP_2 > 10, TRUE, FALSE),
    endocrine = if_else(UMAP_1 > -1 & UMAP_1 < -0.2 & UMAP_2 > 10, TRUE, FALSE),
    unkown = if_else(UMAP_1 > -6 & UMAP_1 < -4.5 & UMAP_2 > 6 & UMAP_2 < 8, TRUE, FALSE),
    nk_t = if_else(UMAP_1 > 10 & UMAP_2 < 0, TRUE, FALSE)
    )

pDCs_id <- annot_23[annot_23$pDCs, "ID"]
endothelial_id <- annot_23[annot_23$endothelial, "ID"]
endocrine_id <- annot_23[annot_23$endocrine, "ID"]
nk_t_id <- annot_23[annot_23$nk_t, "ID"] # These fall within clusters 24 and 29
unk_id <- annot_23[annot_23$unkown, "ID"]
```

Now to get the markers of the unkown cluster
```{r}
pdac_A[["preannot"]] <- if_else(annot_23$unkown, "unkown", as.character(annot_23$SCT_snn_res.5))

Seurat::Idents(object = pdac_A) <- pdac_A@meta.data[, "preannot"]
markers_pdacA_update <- Seurat::FindAllMarkers(object = pdac_A,
                                        verbose = TRUE,
                                        only.pos = TRUE,
                                        assay = "SCT",
                                        slot = "data")

markers_pdacA_update %>% filter(cluster == "unkown")
```


From looking at marker genes and matchscore we can get the following annotation:
```{r}
hypox <- "Hypoxia Ductal Cells"
centroacinar <- "Centroacinar Ductal Cells"
terminal <- "Terminal Ductal Cells"
antigen <- "Antigen-presenting ductal cells"
cancer1 <- "Malignant Ductal Cells (TM4SF1)"
cancer2 <- "Malignant Ductal Cells (S100A4)"
macrophage <- "Macrophage"
mast <- "MAST"
fibroblast <- "Fibroblasts cancer"
nk <- "NK"
tc <- "T cells"
tc_nk <- paste(nk, tc, sep = "/")
acinar <- "Acinar Cells"
islet <- "Pancreatic islet"
dc <- "DC"
pdc <- "pDCs"
endot <- "Endothelial"
endoc <- "Endocrine"
rbc <- "RBC"

annot_ds <- data.frame(SCT_snn_res.5 = as.factor(0:31), 
           annotation = c(hypox, 
                          centroacinar, 
                          cancer1, 
                          terminal, 
                          centroacinar, 
                          terminal, 
                          antigen, 
                          cancer2, 
                          antigen,
                          antigen,
                          antigen,
                          centroacinar,
                          centroacinar,
                          centroacinar,
                          terminal,
                          antigen,
                          macrophage,
                          macrophage,
                          centroacinar,
                          terminal,
                          islet,
                          cancer2,
                          hypox,
                          hypox,
                          tc_nk,
                          cancer2,
                          cancer2,
                          centroacinar,
                          mast,
                          tc_nk,
                          hypox,
                          acinar))

metadata_embed <- pdac_A@meta.data %>%
  tibble::rownames_to_column("ID") %>% 
  dplyr::left_join(annot_ds, by = "SCT_snn_res.5") %>% 
  dplyr::left_join(data.frame(pdac_A@reductions$umap@cell.embeddings) %>% rownames_to_column("ID"), by = "ID") %>% 
  dplyr::mutate(
    annotation = if_else(ID %in% nk_t_id, tc_nk, as.character(annotation)),
    annotation = if_else(ID %in% pDCs_id, pdc, as.character(annotation)),
    annotation = if_else(ID %in% endothelial_id, endot, as.character(annotation)),
    annotation = if_else(ID %in% endocrine_id, endoc, as.character(annotation)),
    annotation = if_else(ID %in% unk_id, rbc, as.character(annotation)))

pdac_A[["annotation"]] <- metadata_embed$annotation

DimPlot(pdac_A, group.by = "annotation")
```

### Save annotaded object
```{r}
saveRDS(object = pdac_A,
        file = "analysis/pancreas_PDAC/data/pdac_A_annot.rds")
```


#### pdac_B
```{r}
markers_pdacB_ls <- lapply(unique(markers_pdacB$cluster), function(i) {
  markers_pdacB[markers_pdacB$cluster == i, "gene"]
})

ms <- matchSCore2::matchSCore2(gene_cl.ref = marker_ls,
                               gene_cl.obs = markers_pdacB_ls, 
                               ylab = "gene_list", 
                               xlab = "clusters")
ms
```

Together with matchsocre we are going to look at gene expression in the clusters
```{r}
ref_plt <- DimPlot(pdac_B, reduction = "umap", group.by = "SCT_snn_res.5", label = TRUE)
HoverLocator(plot = ref_plt)
Seurat::FeaturePlot(object = pdac_B,
                    features = c("TM4SF1", # Cancer subtype marker mutually exclusive with S100A4
                                 "S100A4", # Cancer subtype marker mutually exclusive with TM4SF1
                                 "KRT19" # Ubiquitous in malignant and non-malignant ductal cells 
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_B,
                    features = c("APOL1", # Hypoxic ductal subpopulation 1
                                 "ERO1A", # Hypoxic ductal subpopulation 1
                                 "CA9" # Hypoxic ductal subpopulation 1
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_B,
                    features = c("TFF1", # Terminal ductal subpopulation
                                 "TFF2", # Terminal ductal subpopulation
                                 "TFF3" # Terminal ductal subpopulation
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_B,
                    features = c("CRISP3", # centroacinar ductal population
                                 "CFTR", # centroacinar ductal population
                                 "TFF1",
                                 "AQP3"
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_B,
                    features = c("CD74", # antigen-presenting ductal cells
                                 "HLA-DPA1", # antigen-presenting ductal cells
                                 "HLA-DQA2", # antigen-presenting ductal cells
                                 "HLA-DRA", # antigen-presenting ductal cells
                                 "HLA-DRB1", # antigen-presenting ductal cells
                                 "C1S", "C4A", "C4B", "CFB", "CFH"
                                 )) + ref_plt

Seurat::FeaturePlot(object = pdac_B,
                    features = c("IL1B", # MacrophagesI, inflammatory M1 stat
                                 "CD163", # MacrophagesII alternatively activated
                                 "MS4A4A", # MacrophagesII alternatively activated
                                 "CD68", # General Myeloyd marker
                                 "C1QA" # Macrophage specific
                                 )) + ref_plt

### MAST ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("CD117", "CD203c", "CD25", "KIT", "SLC18A2")) + ref_plt

### Monocyte ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("CD14", "CCR2", "CCR7", "CD68")) + ref_plt

### DC ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("CD14", "CD40", "CLEC4C")) + ref_plt
# "CD11c", "CD135", "CD141", "CD172a", "CD197", "CD1b", "CD1c", "CD205", "CD206",  "CD273", "CD282",  "CD284", "CD289", "CD369", "CD370", "CD49d", "CD207"

### pDCs ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("CD123", "GZMB", "BDCA2", "NRP1", "CLEC4C")) + ref_plt

### NK ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = marker_ls[["NK"]]) + ref_plt

### CD8 T cells ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = marker_ls[["CD8+ T cells"]]) + ref_plt

### CD4 T cells ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = marker_ls[["CD4+ T cells"]]) + ref_plt

### Pancreatic islet cells ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c(marker_ls[["pancreatic islet"]][1:10], "AZGP1")) + ref_plt

### Endothelial ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("CD31", "CD34", "CD146", "CD105", "VWF")) + ref_plt

### Epithelial ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("THY1")) + ref_plt

### Endocrine ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("ALAS2", "HBA1", "HBB", "HBG1")) + ref_plt

### Erythrocytes ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("CD235b", "CD45", "BCL2", "IAP", "HBB")) + ref_plt

### tufts cells ###
Seurat::FeaturePlot(object = pdac_B, 
                    features = c("ACADSB", "ACSL4", "ADA", "ADAM10", "AEN", 
                                 "AFAP1", "AHCY", "ALPL", "ANXA1", "ANXA4", 
                                 "ASCL1", "ASCL2", "ATP11C", "AZGP1")) + ref_plt

DimPlot(pdac_B, reduction = "umap", group.by = "SCT_snn_res.5", label = TRUE) + theme_bw()
```

Check difference between the 2 ductal cell clusters
```{r}
pdac_B[["broad"]] <- if_else(pdac_B$SCT_snn_res.5 %in% c(0,2,11,14,32,36), 
                             "Ductal-2", if_else(pdac_B$SCT_snn_res.5 %in% c(3,4,5,6,7,9,10,12,17,18,19,20,21,23,27),
                                     "Ductal-1", as.character(pdac_B$SCT_snn_res.5)))

Seurat::Idents(object = pdac_B) <- pdac_B@meta.data[, "broad"]
mrk_ls <- FindMarkers(object = pdac_B, ident.1 = "Ductal-1", ident.2 = "Ductal-2")
DimPlot(pdac_B, reduction = "umap", group.by = "broad", label = TRUE) + theme_bw()

mrk_pos <- mrk_ls %>% 
  rownames_to_column("gene") %>% 
  filter(p_val_adj < 0.01 & avg_logFC > 1)

mrk_neg <- mrk_ls %>% 
  rownames_to_column("gene") %>% 
  filter(p_val_adj < 0.01 & avg_logFC < 1)

```

Look at the unannotated clusters
```{r}
mast_sel <- pdac_B@meta.data %>% 
  data.frame() %>% 
  tibble::rownames_to_column("ID") %>% 
  dplyr::left_join(data.frame(pdac_B@reductions$umap@cell.embeddings) %>% rownames_to_column("ID"), by = "ID") %>%
  dplyr::mutate(
    MAST = if_else(UMAP_1 > 0 & UMAP_2 > 15, TRUE, FALSE)) %>%
  dplyr::filter(MAST == TRUE) %>% 
  dplyr::pull(ID)
```

From looking at marker genes and matchscore we can get the following annotation:
```{r}
hypox <- "Hypoxia Ductal Cells"
centroacinar <- "Centroacinar Ductal Cells"
terminal <- "Terminal Ductal Cells"
antigen <- "Antigen-presenting ductal cells"
cancer1 <- "Malignant Ductal Cells (TM4SF1)"
cancer2 <- "Malignant Ductal Cells (S100A4)"
macrophage <- "Macrophage"
mast <- "MAST"
fibroblast <- "Fibroblasts cancer"
nk <- "NK"
tc <- "T cells"
acinar <- "Acinar Cells"
islet <- "Pancreatic islet"
dc <- "DC"
pdc <- "pDCs"
endot <- "Endothelial"
endoc <- "Endocrine"
neutro <- "Neutrophil"
mono <- "Monocytes"
epith <- "Epithelial"


annot_ds <- data.frame(SCT_snn_res.5 = as.factor(0:36), 
           annotation = c(terminal, 
                          cancer1, 
                          terminal, 
                          antigen, 
                          terminal, 
                          terminal, 
                          terminal, 
                          centroacinar, 
                          endot,
                          centroacinar,
                          terminal,
                          terminal,
                          terminal,
                          cancer1,
                          terminal,
                          cancer1,
                          cancer1,
                          terminal,
                          terminal,
                          terminal,
                          terminal,
                          terminal,
                          endot,
                          terminal,
                          terminal,
                          macrophage,
                          cancer1,
                          terminal,
                          islet,
                          neutro,
                          tc,
                          endot,
                          terminal,
                          dc,
                          acinar,
                          epith,
                          terminal))

metadata_embed <- pdac_B@meta.data %>%
  tibble::rownames_to_column("ID") %>% 
  dplyr::left_join(annot_ds, by = "SCT_snn_res.5") %>% 
  dplyr::left_join(data.frame(pdac_B@reductions$umap@cell.embeddings) %>% rownames_to_column("ID"), by = "ID") %>% 
  dplyr::mutate(
    annotation = if_else(ID %in% mast_sel, mast, as.character(annotation)))

pdac_B[["annotation"]] <- metadata_embed$annotation

DimPlot(pdac_B, group.by = "annotation", label = TRUE)
DimPlot(pdac_B, label = TRUE)
```

Remove low quality cell cluster, we can see that cluster made out of subclusters 0, 2, 11, 14, 24, 32, and 36 have a high mitochondrial content and low UMI count.
```{r}
FeaturePlot(object = pdac_B, 
            features = c("percent.mt", "nFeature_RNA"))

pdac_B <- pdac_B[, !pdac_B$SCT_snn_res.5 %in% c(0, 2, 11, 14, 24, 32, 36)]
```



### Save annotaded object
```{r}
saveRDS(object = pdac_B,
        file = "analysis/pancreas_PDAC/data/pdac_B_annot.rds")
```

